test/classes/diten/breakList.good

```diff
@@ -1,7 +1,2 @@
-breakList.chpl:23: In function 'bar':
-breakList.chpl:26: error: multiple overload sets are applicable to this call
-breakList.chpl:6: note: the best-matching candidate is here
-breakList.chpl:3: note: ... defined in this module
-$CHPL_HOME/modules/packages/LinkedLists.chpl:nnnn: note: even though the candidate here is also available
-$CHPL_HOME/modules/packages/LinkedLists.chpl:nnnn: note: ... defined in this module
-breakList.chpl:26: note: use --no-overload-sets-checks to disable overload sets errors
+3.0 4.0
+breakList.chpl:7: error: halt reached - I Got you!
```

Comment:
This test goes from a multiple overload sets error to running and printing "I Got you!". That is because the LibGL shadow scope is closer than the LinkedLists shadow scope. I think this had the previous behavior only due to bugs in isMoreVisible in function resolution.

Reply:
Making sure we're on the same page with the behavior in the new .good here:  The lesson here is "When you rely on unqualified `use` and don't pay attention to what you're actually getting from it, you may get bit and there's nothing we can do about that, right?

But I'm also confused:  What makes the "got you" `.append()` overload preferable to the original?  My intuition would be that there'd be an ambiguity rather than that this one would win.  Let's see... why would I think that...?  I suppose because I don't think of methods as being closer to or further from the callsite relative to each other, but all visible methods associated with a given type as being equally available since they're accessed through the type rather than scoping (as with a standalone routine or variable)?

Reply:
Michael brought up some questions about method visibility in #19352, it'd probably be good to take a look over there

Reply:
It comes down to this:

``` chapel
  use LinkedLists;
  proc bar() {
    use LibGL; // defines LinkedList.append
    var li = new LinkedList(real);
    li.append(3); // uses LibGL.LinkedList.append because the 'use' had closer scope
  }
```

We could make such a case an ambiguity error but we also use it a lot (e.g. to write those compatibility functions to adapt to different Chapel versions). The problem doesn't strictly go away with `import` either if you use `import LibGL.LinkedList` to bring in those methods. But then, is it surprising to run LibGL.LinkedList.append? Probably not. OTOH, if you just had `import LibGL` then there is no place for hijacking here.

So yeah, `use` is overly broad, and `import SomeModule.SomeType` can potentially be more broad than you would like (I'm not sure how to solve that but #19352 discusses it).

Reply:
> uses LibGL.LinkedList.append because the 'use' had closer scope

Yeah, this is where I was trying to say that—for whatever reason—when it comes to methods, I think of them as not being as subject to "having closer scope" as standalone stuff, particularly after our recent primary/secondary/tertiary discussions.  Instead, I sortof imagine I can see a mass of methods through various uses/imports that bring them in, and all are equally close to me because I'm getting to them through the `this` rather than through scoping.  I realize this is not something we've discussed before, it's just where I realize my head goes when looking at this example.

> we also use it a lot (e.g. to write those compatibility functions to adapt to different Chapel versions).

Can you refer me to an example?  I think most of the compatibility stuff I've done has been on standalone declarations rather than methods.

Reply:
>  Instead, I sortof imagine I can see a mass of methods through various uses/imports that bring them in, and all are equally close to me because I'm getting to them through the this rather than through scoping.

This is an interesting viewpoint & we should discuss in more in #19352.

Reply:
> > we also use it a lot (e.g. to write those compatibility functions to adapt to different Chapel versions).
> 
> Can you refer me to an example?  I think most of the compatibility stuff I've done has been on standalone declarations rather than methods.

I may be wrong about its prevalence but the top post in #17438 has something of this nature (not necessarily for methods).

Reply:
> This is an interesting viewpoint & we should discuss in more in https://github.com/chapel-lang/chapel/issues/19352.

I dropped a version of this viewpoint there.

Reply:
Please see https://github.com/chapel-lang/chapel/issues/19352#issuecomment-1079391291 about this test

Reply:
My takeaway from a quick review of #19352 is that we'd prefer to make this an ambiguity error, but it looks as though we haven't done that yet.  Am I correct on both counts?  What's the plan for this one?

Reply:
Please see https://github.com/chapel-lang/chapel/issues/19352#issuecomment-1079391291 - this test behaves this way due to the argument types of `x` - `int` vs `real`. There remains an open question about what overload set checking should do for methods.

Right now, on the overloads set checking question, this PR does consider "more visible" when doing overload set checking between methods (it does not do that when disambiguating). This is the reason there is not an overloads set checking error.

I think the current situation is reasonable for this PR / design proposal and that the question of whether overload set checking should consider isMoreVisible for methods is a separate question we can and should answer later.